cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh

```diff
@@ -0,0 +1,488 @@
+    const int want_bucket        = counter->bucket;
+    IdxT& filter_cnt             = counter->filter_cnt;
+    IdxT& out_cnt                = counter->out_cnt;
+    T& kth_value                 = counter->kth_value;
+    const IdxT counter_len       = counter->len;
+    const int previous_start_bit = calc_start_bit<T, BITS_PER_PASS>(pass - 1);
+    const unsigned previous_mask = calc_mask<T, BITS_PER_PASS>(pass - 1);
+
+    auto f = [in_idx_buf,
+              out_buf,
+              out_idx_buf,
+              out,
+              out_idx,
+              greater,
+              k,
+              start_bit,
+              mask,
+              previous_start_bit,
+              previous_mask,
+              want_bucket,
+              &filter_cnt,
+              &out_cnt,
+              &kth_value,
+              counter_len](T value, IdxT i) {
+      int prev_bucket =
+        calc_bucket<T, BITS_PER_PASS>(value, previous_start_bit, previous_mask, greater);
+      if (prev_bucket == want_bucket) {
+        IdxT pos     = atomicAdd(&filter_cnt, IdxT(1));
+        out_buf[pos] = value;
+        if (out_idx_buf) { out_idx_buf[pos] = in_idx_buf ? in_idx_buf[i] : i; }
+        int bucket = calc_bucket<T, BITS_PER_PASS>(value, start_bit, mask, greater);
+        atomicAdd(histogram_smem + bucket, IdxT(1));
+
+        if (counter_len == 1) {
+          if (out) {
+            out[k - 1]     = value;
+            out_idx[k - 1] = in_idx_buf ? in_idx_buf[i] : i;
+          } else {
+            kth_value = value;
+          }
+        }
+      } else if (out && prev_bucket < want_bucket) {
+        IdxT pos     = atomicAdd(&out_cnt, IdxT(1));
+        out[pos]     = value;
+        out_idx[pos] = in_idx_buf ? in_idx_buf[i] : i;
+      }
+    };
+
+    vectorized_process(in_buf, previous_len, f);
+  }
+  __syncthreads();
+
+  for (int i = threadIdx.x; i < num_buckets; i += blockDim.x) {
+    if (histogram_smem[i] != 0) { atomicAdd(histogram + i, histogram_smem[i]); }
+  }
+}
+
+template <typename IdxT, int BITS_PER_PASS, int NUM_THREAD>
+__device__ void scan(volatile IdxT* histogram,
+                     const int start,
+                     const int num_buckets,
+                     const IdxT current)
+{
+  typedef cub::BlockScan<IdxT, NUM_THREAD> BlockScan;
+  __shared__ typename BlockScan::TempStorage temp_storage;
+
+  IdxT thread_data = 0;
+  int index        = start + threadIdx.x;
+  if (index < num_buckets) { thread_data = histogram[index]; }
+
+  BlockScan(temp_storage).InclusiveSum(thread_data, thread_data);
+  __syncthreads();
+  if (index < num_buckets) { histogram[index] = thread_data + current; }
+  __syncthreads();  // This sync is necessary, as the content of histogram needs
+                    // to be read after
+}
+
+template <typename T, typename IdxT, int BITS_PER_PASS, int NUM_THREAD>
+__device__ void choose_bucket(Counter<T, IdxT>* counter, IdxT* histogram, const IdxT k)
```

Comment:
update of the counter:
The k-th value must be in the `index`-th bucket, so we record it in `counter->bucket`.
The number of values in this bucket is `(cur - prev)` since `histogram` now actually contains prefix sum of the histogram. Hence the update of `counter->len`.
counter is used in the next pass, so we keep the len of current pass in `counter->previous_len`.

`prev` means the number of values in the first `index-1` buckets. These values are less than (when `greater`=false) the real k-th value, so must be in the top-k results. In the next pass, we need only to find the `(k - prev)`-th value fall in the `index`-th bucket. Hence the setting of `counter->k`.